syncd/syncd.cpp

```diff
@@ -2954,25 +2957,98 @@ void processFlexCounterGroupEvent(
     }
 }
 
+std::queue<swss::KeyOpFieldsValuesTuple> g_flexCounterEventQueue;
+
+bool tryPopFlexCounterEvent(
+        _Out_ swss::KeyOpFieldsValuesTuple& kco)
+{
+    SWSS_LOG_ENTER();
+
+    std::lock_guard<std::mutex> lock(g_mutex);
+
+    if (g_flexCounterEventQueue.empty())
+    {
+        return false;
+    }
+
+    kco = g_flexCounterEventQueue.front();
+
+    g_flexCounterEventQueue.pop();
+
+    return true;
+}
+
+void pushFlexCounterEvent(
+    _In_ const swss::KeyOpFieldsValuesTuple& kco)
+{
+    SWSS_LOG_ENTER();
+
+    std::lock_guard<std::mutex> lock(g_mutex);
+
+    g_flexCounterEventQueue.push(kco);
+}
+
+bool processFlexCounterEvent(
+    _In_ const swss::KeyOpFieldsValuesTuple kco);
+
+void processFlexCounterEventThread()
+{
+    SWSS_LOG_ENTER();
+
+    while (g_processFlexCounterEventThreadRun)
+    {
+        swss::KeyOpFieldsValuesTuple kco;
+
+        if (tryPopFlexCounterEvent(kco))
+        {
+            if (!processFlexCounterEvent(kco))
+            {
+                // event was not successfully processed, put it again to the queue
+
+                pushFlexCounterEvent(kco);
```

Comment:
This may upset the event sequence in the processing failure case. The processing failure case, however, is the problem you would like to solve.

For the flex counter table in particular, it typically contains the counter id list that is expected to remain static and unchanged after SET. We should not see any two operations on the same key in the queue at the same time, although there leaves the possibility of running into such a sequence problem.

If we enqueue at the front and retry, we can avoid the event sequence problem. However, we may experience head of line blocking until the event processing retry succeeds.

Reply:
i don't get the problem you described here, all those issues with counters are present here which we want to workaround because they are executed on the same channel and can be picked up by select  independently, and i stated at the beginning that subscribe to the counter is broken design and of course you can came up with scenario that can break even this workaround, since you need to wait for RIF to be created and for example potentially not destroyed and created again, which all would be solved when syncd would become synchronous on which i already send  PR